src/api/crypto/blockchain/Imp_blockchain.cpp

```diff
@@ -381,72 +409,82 @@ auto BlockchainImp::ProcessMergedContact(
     return true;
 }
 
-auto BlockchainImp::ProcessTransaction(
+auto BlockchainImp::ProcessTransactions(
     const opentxs::blockchain::Type chain,
-    const opentxs::blockchain::block::bitcoin::Transaction& in,
+    Set<std::shared_ptr<opentxs::blockchain::bitcoin::block::Transaction>>&& in,
     const PasswordPrompt& reason) const noexcept -> bool
 {
     const auto& db = api_.Network().Blockchain().Internal().Database();
-    auto copy = in.clone();
+    const auto& log = LogTrace();
+    auto lock = Lock{lock_};
 
-    OT_ASSERT(copy);
+    for (auto& pTX : in) {
+        OT_ASSERT(pTX);
 
-    auto lock = Lock{lock_};
-    const auto& id = in.ID();
-    const auto txid = id.Bytes();
+        const auto& tx = *pTX;
+        const auto& id = tx.ID();
+        const auto txid = id.Bytes();
+        auto old = db.LoadTransaction(txid);
+        auto proto = proto::BlockchainTransaction{};
 
-    if (auto tx = db.LoadTransaction(txid); tx) {
-        tx->Internal().MergeMetadata(chain, in.Internal(), LogTrace());
+        if (old) {
+            old->Internal().MergeMetadata(chain, tx.Internal(), log);
 
-        if (false == db.StoreTransaction(*tx)) {
-            LogError()(OT_PRETTY_CLASS())(
-                "failed to save updated transaction ")(id.asHex())
-                .Flush();
+            if (false == db.StoreTransaction(*old, proto)) {
+                LogError()(OT_PRETTY_CLASS())(
+                    "failed to save updated transaction ")(id.asHex())
+                    .Flush();
 
-            return false;
+                return false;
+            }
+        } else {
+            if (false == db.StoreTransaction(tx, proto)) {
+                LogError()(OT_PRETTY_CLASS())(
+                    "failed to save new transaction ")(id.asHex())
+                    .Flush();
+
+                return false;
+            }
         }
 
-        copy.swap(tx);
-    } else {
-        if (false == db.StoreTransaction(in)) {
-            LogError()(OT_PRETTY_CLASS())("failed to save new transaction ")(
-                id.asHex())
+        if (false == db.AssociateTransaction(id, tx.Internal().GetPatterns())) {
+            LogError()(OT_PRETTY_CLASS())(
+                "associate patterns for transaction ")(id.asHex())
                 .Flush();
 
             return false;
         }
-    }
 
-    if (false == db.AssociateTransaction(id, copy->Internal().GetPatterns())) {
-        LogError()(OT_PRETTY_CLASS())("associate patterns for transaction ")(
-            id.asHex())
-            .Flush();
+        if (!reconcile_activity_threads(lock, proto, (old ? *old : tx))) {
 
-        return false;
+            return false;
+        }
     }
```

Comment:
this condition tree can at least partially be closed as a result of logical operations

Reply:
MT-82